File: app/src/main/java/com/castelijns/mmdemo/users/UsersRepo.java
package com.castelijns.mmdemo.users;

import com.castelijns.mmdemo.models.User;
import com.castelijns.mmdemo.network.ApiManager;
import com.castelijns.mmdemo.network.ApiService;

import java.util.List;

import io.reactivex.Observable;

public class UsersRepo {

    private static UsersRepo instance = null;

    private static ApiService apiService;

    private UsersRepo() {
        apiService = ApiManager.getService();
    }

    public static UsersRepo getInstance() {
        if (instance == null) {
            instance = new UsersRepo();
        }

        return instance;
    }

    public Observable<List<User>> getAllUsers() {
        return apiService.getAllUsers();
    }
}

File: app/src/main/kotlin/com/castelijns/mmdemo/users/UsersRepo.kt
package com.castelijns.mmdemo.users

import com.castelijns.mmdemo.app.SimpleCache
import com.castelijns.mmdemo.models.User
import com.castelijns.mmdemo.network.ApiManager
import kotlinx.coroutines.experimental.Deferred
import kotlinx.coroutines.experimental.Dispatchers
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.coroutineScope

object UsersRepo : SimpleCache<User>() {

    private var apiService = ApiManager.service

    suspend fun getAllUsers(): Deferred<List<User>> = coroutineScope {
        async(Dispatchers.IO) {
            cache ?: apiService.getAllUsers().await()
        }
    }
}
